Pin the versions of the dependencies.

clalancette opened the pull request:
What this does is to update our dockerfile to use Jammy, and
then install as many dependencies as possible via the system
packages.  We then take the versions of those dependencies and
pin them in requirements.txt to those versions.  This is all
so that we have a consistent set of dependencies that we know
will work for the lifetime of Ubuntu Jammy, without having
to constantly monitor the upstream projects.  We'll update this
again the next time we switch Ubuntu distributions.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Pin the versions of the dependencies.

This installs as many dependencies as possible via the system
packages.  We then take the versions of those dependencies
and pin them in requirements.txt/constraints.txt .  This is
so that we keep a consistent set of dependencies that we know
will work for the lifetime of Ubuntu Jammy, without constantly
monitoring the upstream projects.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

audrow reviewed (approved):
I think this is a good idea, and looks good to me 👍

nuclearsandwich commented:
100% here for the idea. I think that implementation wise it would be better to use a [constraints file](https://pip.pypa.io/en/latest/user_guide/#constraints-files) to pin versions not just of our dependencies but their transitive dependencies as well. I will add patches to this PR with my proposed implementation.

clalancette commented:
@audrow @nuclearsandwich I know this is an old PR, but I'd like to move forward here and try to get this in.  I've just rebased it and added in the constraints file as suggested earlier on.  I'd appreciate another review here if possible, since it is fairly different from when I first opened it.  Thanks!

audrow reviewed (approved):
Looks good to me!

clalancette commented:
Thanks!  I'm going to go ahead and merge this, and then deploy it to make sure things work as we expect.
